File: exercism/javascript/complete/protein-translation/protein-translation.js
const protein_codons = {
  "Methionine":["AUG"],
  "Phenylalanine":["UUU","UUC"],
  "Leucine":["UUA","UUG"],
  "Serine":["UCU","UCC","UCA","UCG"],
  "Tyrosine":["UAU","UAC"],
  "Cysteine":["UGU","UGC"],
  "Tryptophan":["UGG"],
  "STOP":["UAA","UAG","UGA"]
}

const translate = (rna) => {
  let transcriptase = [];
  if (rna == undefined) return transcriptase;
  if (rna.match(/[^ACGU]/)) throw (new Error('Invalid codon'));
  for(let i = 0; i < rna.length; i += 3) {
    let codon = rna[i] + rna[i+1] + rna[i+2];
    for (let protein in protein_codons) {
      protein_codons[protein].forEach((triplets) => {
        if (codon == triplets && protein == "STOP") {
          i += rna.length;
        } else if (codon == triplets) {
          transcriptase.push(protein);
        }
      })
    }
  }
  return transcriptase;
}

export default translate;

File: exercism/ruby/complete/protein-translation/protein_translation.rb
class InvalidCodonError < ArgumentError
end

class Translation
  @@transcriptase = {
    Methionine: %w(AUG),
    Phenylalanine: %w(UUU UUC),
    Leucine: %w(UUA UUG),
    Serine: %w(UCU UCC UCA UCG),
    Tyrosine: %w(UAU UAC),
    Cysteine: %w(UGU UGC),
    Tryptophan: %w(UGG),
    STOP: %w(UAA UAG UGA)
  }

  def self.of_codon(codon)
    @@transcriptase.each{ |key, value| return key.to_s if value.include?(codon) }
    false
  end

  def self.of_rna(strand)
    output = []
    i = 0
    while i < strand.length
      protein = self.of_codon(strand[i...i+3])
      break if protein == "STOP"
      raise InvalidCodonError unless protein
      output << protein
      i += 3
    end
    output
  end
end
